jetty-websocket/websocket-javax-common/src/main/java/org/eclipse/jetty/websocket/javax/common/decoders/AvailableDecoders.java

```diff
@@ -200,107 +133,83 @@ private void add(Class<? extends Decoder> decoder, Class<? extends Decoder> inte
         if (objectType == null)
         {
-            StringBuilder err = new StringBuilder();
-            err.append("Unknown Decoder Object type declared for interface ");
-            err.append(interfaceClass.getName());
-            err.append(" on class ");
-            err.append(decoder);
-            throw new InvalidWebSocketException(err.toString());
+            String err = "Unknown Decoder Object type declared for interface " +
+                interfaceClass.getName() + " on class " + decoder;
+            throw new InvalidWebSocketException(err);
         }
 
-        try
+        // Validate the decoder to be added against the existing registered decoders.
+        for (RegisteredDecoder registered : registeredDecoders)
         {
-            RegisteredDecoder conflicts = registeredDecoders.stream()
-                .filter(registered -> registered.isType(objectType))
-                .filter(registered -> !registered.primitive)
-                .findFirst()
-                .get();
-
-            if (conflicts.decoder.equals(decoder) && conflicts.implementsInterface(interfaceClass))
+            if (!registered.primitive && objectType.equals(registered.objectType))
             {
-                // Same decoder as what is there already, don't bother adding it again.
-                return;
+                // Streaming decoders can only have one decoder per object type.
+                if (interfaceClass.equals(Decoder.TextStream.class) || interfaceClass.equals(Decoder.BinaryStream.class))
+                    throw new InvalidWebSocketException("Multiple decoders for objectType" + objectType);
+
+                // If we have the same objectType, then the interfaceTypes must be the same to form a decoder list.
+                if (!registered.interfaceType.equals(interfaceClass))
+                    throw new InvalidWebSocketException("Multiple decoders with different interface types for objectType " + objectType);
             }
 
-            StringBuilder err = new StringBuilder();
-            err.append("Duplicate Decoder Object type ");
-            err.append(objectType.getName());
-            err.append(" in ");
-            err.append(decoder.getName());
-            err.append(", previously declared in ");
-            err.append(conflicts.decoder.getName());
-            throw new InvalidWebSocketException(err.toString());
-        }
-        catch (NoSuchElementException e)
-        {
-            registeredDecoders.addFirst(new RegisteredDecoder(decoder, interfaceClass, objectType));
+            // If this decoder is already registered for this interface type we can skip adding a duplicate.
+            if (registered.decoder.equals(decoder) && registered.interfaceType.equals(interfaceClass))
+                return;
         }
+
+        registeredDecoders.add(0, new RegisteredDecoder(decoder, interfaceClass, objectType, config));
```

Comment:
Please add a comment about why it is added at index `0`.
In this way the decoder order is reversed, is that ok?
That is, if 2 decoders are added, the order will be reversed which may be unexpected, no?

Reply:
This is so any decoders added by the endpoint take precedence over the primitive decoders. The decoders supplied by the endpoint config are given as a list:
```java
ClientEndpointConfig.Builder.create()
    .decoders(List.of(Decoder1.class, Decoder2.class))
    .build();
```

So in this case should `Decoder1` be tried before `Decoder2`? I don't think the spec goes into this much detail about it.

Reply:
Looks like I can fix the precedence issue by just registering the configured decoders before the primitive ones, then I can avoid adding at index `0` here.

Reply:
I think the order should be preserved, to minimize surprises to applications.